consensus/src/liveness/proposal_generator.rs

```diff
@@ -156,53 +157,66 @@ impl PipelineBackpressureConfig {
     }
 
-    pub fn get_execution_block_size_backoff(
+    fn compute_lookback_blocks(
+        &self,
+        block_execution_times: &[ExecutionSummary],
+        f: impl Fn(&ExecutionSummary) -> Option<u64>,
+    ) -> Vec<u64> {
+        block_execution_times
+            .iter()
+            .flat_map(|summary| {
+                // for each block, compute target (re-calibrated) block size
+                f(summary)
+            })
+            .sorted()
+            .collect::<Vec<_>>()
+    }
+
+    fn compute_lookback_metric(&self, blocks: &[u64], metric: &ExecutionBackpressureMetric) -> u64 {
+        match metric {
+            ExecutionBackpressureMetric::Mean => {
+                if blocks.is_empty() {
+                    return 0;
+                }
+                (blocks.iter().sum::<u64>() as f64 / blocks.len() as f64) as u64
+            },
+            ExecutionBackpressureMetric::Percentile(percentile) => *blocks
+                .get(((percentile * blocks.len() as f64) as usize).min(blocks.len() - 1))
+                .expect("guaranteed to be within vector size"),
+        }
+    }
+
+    fn get_execution_block_size_backoff(
         &self,
         block_execution_times: &[ExecutionSummary],
         max_block_txns: u64,
     ) -> Option<u64> {
         self.execution.as_ref().and_then(|config| {
-            let sizes = block_execution_times
-                .iter()
-                .flat_map(|summary| {
-                    // for each block, compute target (re-calibrated) block size
-
-                    let execution_time_ms = summary.execution_time.as_millis();
-                    // Only block above the time threshold are considered giving enough signal to support calibration
-                    // so we filter out shorter locks
-                    if execution_time_ms > config.min_block_time_ms_to_activate as u128
-                        && summary.payload_len > 0
-                    {
-                        // TODO: After cost of "retries" is reduced with execution pool, we
-                        // should be computing block gas limit here, simply as:
-                        // `config.target_block_time_ms / execution_time_ms * gas_consumed_by_block``
-                        //
-                        // Until then, we need to compute wanted block size to create.
-                        // Unfortunatelly, there is multiple layers where transactions are filtered.
-                        // After deduping/reordering logic is applied, max_txns_to_execute limits the transactions
-                        // passed to executor (`summary.payload_len` here), and then some are discarded for various
-                        // reasons, which we approximate are cheaply ignored.
-                        // For the rest, only `summary.to_commit` fraction of `summary.to_commit + summary.to_retry`
-                        // was executed. And so assuming same discard rate, we scale `summary.payload_len` with it.
```

Comment:
this part of the comment still seems relevant for computation here, and why we don't prefer txn_limit.